MITgcmutil update

dramauh opened the pull request:
## What changes does this PR introduce?
I have updated the python package MITgcmutils including new modules


## What is the current behaviour? 
The current version lacks of basic python functions to open and manipulate binary output from MITgcm


## What is the new behaviour 
(if this is a feature change)?
The updated library has converted MATLAB functions such gen_blanklist, readbin and writebin. It also includes all the density profiles in MITgcm, and a function to convert depth to dbar following MITgcm. Finally, there are some new functions in utils.py for tile visualization and/or debugging.

## Does this PR introduce a breaking change? 
(What changes might users need to make in their application due to this PR?)
The documentation needs a slight update to include the new functions and the changes in the old ones

## Other information:
I wrote a symbolic version of 0.5. I let the developers decide the most suitable version number

## Suggested addition to `tag-index`
(To avoid unnecessary merge conflicts, please don't update `tag-index`. One of the admins will do that when merging your pull request.)

mjlosch commented:
When I include `MITgcmutils.utils` in `doc/utilities/utilities.rst` I get error messages like:
```
/Users/mlosch/MITgcm/review/utils/python/MITgcmutils/MITgcmutils/utils.py:docstring of MITgcmutils.utils.tilecmap:30: CRITICAL: Unexpected section title.

Test
----
```
This is from the doc in the python function and seems to be triggered by the preceding 
```
    Example
    -------
    >>> [fig]=tilemap(bathy, 51, 51,)
    >>> [fig]=tilemap(bathy, 51, 51,58,5)
```
When I move `Test` before `Example` the errors go away. I have no idea how to fix this.

mjlosch commented:
My previous (first) comment got scrapped, so I repeat what I can remember of it: @dramauh Thanks for this contribution. I took the liberty to make small changes to make the doc-work. Not sure if I did it correctly ( @jahn ?). Please check if it works for you and please add missing documentation. When I tried, I ran into the problem described above.

dramauh commented:
@mjlosch is fine, I should have tried to compile the document before commiting. That error catched me by surprised  Thank for correcting the problem

dramauh commented:
I fixed a small flag in the density.py. I also corrected documentation adding utils and tests. I did not get any error @mjlosch, and I wrote exactly the same code that you use..very strange

mjlosch commented:
Thanks, I can now also compile the readthedocs without error messages. It may have to do with the infamous DOS eol characters (^M) that I saw in a few places (e.g., the utils.py file is a DOS file, something you may want to change).

mjlosch commented:
@dramauh I am sorry, I was wrong. The problem described above remains. The docstring sections in utils.gen_blanklist and utils.timecmap starting with "Test" lead to the above error message. I don't know what this is.

mjlosch commented:
The docs were not rendering correctly for both MacOS and Linux (my MacOS implementation gave a warning, see above, the Linux version didn't). Apparently, only one block of code is allowed.

mjlosch commented on utils/python/MITgcmutils/CHANGES.txt:
```diff
@@ -1,3 +1,33 @@
+Version 0.5, 2023-04-17
```
I have no feelings about this., but the next step would be 0.2, wouldn't it?

mjlosch commented on utils/python/MITgcmutils/MITgcmutils/conversion.py:
```diff
@@ -0,0 +1,48 @@
+import numpy as np
+
+# Global constants
+gravity  = -9.810000e+00
+rhoConst =  1.027500e+03
+eosRefP0 =  1.013250e+05
```
These parameters are a matter of choice and as such they should be defaults, but not constants.
One can also imagine making gravity a function of latitude (see UNESCO Tech. Pap. in Mar. Sci., 1983, eq 27):
```
grav = 9.780318 * (1.0 + (5.2788e-3 + 2.36e-5 * sin^2(lat)) * sin^2(lat))
```

dramauh replied:
I agree with you. We could add those constant as default input parameters in the function. Also I have been thinking to remove that minus and move it to the equation..It does not make much sense there. I try to correct it tomorrow.

mjlosch reviewed (approved):
This looks good. I didn't test everything thoroughly, but I did find a small problem in diagnostics.readstats, which I took the liberty to fix. @dramauh please check, if it still works for you. I also changed back the files modes to 644 (rw-r--r--) for all regular files (except the script gluemncbic).

The tests with the large (1MB) bathy_test.bin only work, if one is in the tests directory. Maybe that can be made more foolproof (e.g. by specifying a path to the file?). I am not sure. Maybe @jahn has an opinion about this.

The automodule documentation seems to work now.

dramauh commented:
I added @mjlosch suggestions in conversion and test. Now, you can add latitude to the function pfromz. I also add path to locate the binaries that are used in the test. I am  having a error rendering the docs because of new paths. @mjlosch do you know how to fix this?

mjlosch commented:
@jahn when you find the time, it would be great if you have a quick look at this. I think this is good to merge.

dramauh commented:
I removed the bathy file and included a portion in the test_utils.py to save some space. I wonder if the code looks untidy because of this. @mjlosch may you take a look, please？ I can always change the file back and add a separate data file (~20kb).

mjlosch commented:
I am fine with that, no more large bathymetry file! I would probably for the plots of `timecmap` use default figure size and fonts. On my small screen I cannot see the entire figure without resizing.

dramauh commented:
I deleted the figure and font size properties and fixed a few things

edoddridge commented:
I'm taking a look at this at the moment. I'll have a review by the end of the week. Updates soon.

edoddridge commented on utils/python/MITgcmutils/MITgcmutils/density.py:
```diff
@@ -0,0 +1,1063 @@
+
+    Notes
+    -----
+    AUTHOR:     Martin Losch 2002-08-09  (mlosch@mit.edu)
```
Really? I'm skeptical that @mjlosch wrote this routine in 2002. It looks like this comment has been copy/pasted into the new doc strings.

dramauh replied:
As Martin  said, some of these routines came from Matlab scripts and others from the MITgcm source code. I tried to keep the author comments.

edoddridge replied:
👍
I think that's a great approach in general. 

I queried this one because I wasn't aware that the TEOS-10 routines existed in any format in 2002. But, it's before my time, so I don't actually know.

dramauh replied:
hhahah, I have just noticed it..I will correct those time travel typos

edoddridge replied:
This hasn't been corrected yet.

edoddridge commented on utils/python/MITgcmutils/MITgcmutils/density.py:
```diff
@@ -0,0 +1,1063 @@
+    """
+    Computes in-situ density of sea water
+
+    Density of Sea Water using TEOS 48-polynomial.
```
Change to "TEOS-10"

edoddridge commented on utils/python/MITgcmutils/MITgcmutils/jmd95.py:
```diff
@@ -1,222 +0,0 @@
-#
-# created by mlosch on 2002-08-09
-# converted to python by jahn on 2010-04-29
```
Only half of this provenance has been kept in the new doc string.

edoddridge commented on utils/python/MITgcmutils/MITgcmutils/tests/test_utils.py:
```diff
@@ -0,0 +1,107 @@
+bathy=np.array(bathy).astype('f4').reshape([60,60])
+
+def test_blanklist():
+    """Test blanklist generator
```
As written, I don't think it's possible for these tests to fail under python's `pytest` framework.

We need an `assert` statement or a comparison with blessed output that can fail in order for this to be an automated test.

Once that exists, we can add some code to the automated testing (`.github/workflows/build_testing.yml`) to run them every time.

edoddridge commented on utils/python/MITgcmutils/MITgcmutils/utils.py:
```diff
@@ -0,0 +1,363 @@
+
+
+def hfac(depth,rF,hFacMin=0.3,hFacMinDr=50,htype='C'):
+    """
```
I can't find a test for this function.

dramauh replied:
I add a new example for this function

edoddridge commented on utils/python/MITgcmutils/MITgcmutils/utils.py:
```diff
@@ -0,0 +1,363 @@
+    return tuple(hfac_dic[i] for i in htype)
+
+def readbin(fname, ndims, dataprec='float32', machineformat='b'):
+    """
```
We could have a circular test for the read and write binary functions. Make an array, write it to binary, read it back in, and then assert that the original array is identical to the one that has been through both functions.

dramauh replied:
I did not add it to prevent making dummy files. I prefer to wait for more complex functions before using circular tests.

edoddridge reviewed (changes requested):
These are great additions to the package. I've left a few comments about things to improve or change. I haven't yet had a chance to actually use the new functions on some data, so the comments are, for now, focused on testing frameworks and coding style.

I'll try and make time to point the functions at some data and actually try them out before ocean sciences, but I might not manage it.

dramauh commented on utils/python/MITgcmutils/MITgcmutils/conversion.py:
```diff
@@ -0,0 +1,103 @@
+
+# Global constants
+gravity  =  9.810000e+00
+rad      =  np.pi/180
```
I will take a good look to address all these comments thanks for all your help reviewing the changes

dramauh replied:
I changed deg2rad=np.pi/180 to np.deg2rad

mjlosch commented:
@jahn and I went through the file `density.py` and simplified a few things. The check for dimensions relies on python's ability to broadcast the shapes of different fields and avoids the Matlab-methods inherited from the seawater library. @dramauh Please check if you can live with that. 

I also have a suggestion for the "tests": as they are not unit or regression tests, but rather a useful illustration of `gen_blanklist` and `tilecmap`, I suggest to change the corresponding names to something like "examples".

dramauh commented:
I have just taken a a look to it. I think it is much cleaner now. Thanks for your help. About the tests, I have not preference. I can change the name of those functions. The aim of that folder was to collect all test and/or examples as the library becomes bigger

mjlosch commented:
I think renaming the directory "tests" to "examples" would be good and maybe even rename `test_utils.py`. @dramauh please go ahead an do so.

I am not a good resource for giving advice on how to design automated tests. I just learned something about that yesterday (thanks to @jahn).

mjlosch commented on utils/python/MITgcmutils/MITgcmutils/utils.py:
```diff
@@ -0,0 +1,363 @@
+
+    fig = plt.figure()
+    ax = fig.add_subplot(111)
+    ax.pcolor(mland,cmap=cmap_lm)
```
here and elsewhere: is there any reason to prefer `pooler` over the (in my experience) much faster `pcolormesh`?
I tried `pcolormesh` and the figures look the same.

dramauh replied:
I changed this as you suggested

jahn commented:
I did some more cleanup.  utils.py still needs a bit of attention.

Another issue is that before this PR, one could import MITgcmutils without having matplotlib installed, as long as one didn't explicitly import cs or llc.  Now everything is imported automatically and plotting and non-plotting capabilities are more mixed up.

dramauh commented on doc/utilities/utilities.rst:
```diff
@@ -6,22 +6,27 @@ Utilities
+This Python package includes a number of helpful functions and scripts for
+dealing with MITgcm output.  You can install it from the model repository (in
+directory :filelink:`utils/python/MITgcmutils`) or from the Python Package
+Index:
```
I think this was done by another author. I decided to leave it as it is part of the readocs

edoddridge commented on utils/python/MITgcmutils/MITgcmutils/examples/eg_utils.py:
```diff
@@ -0,0 +1,124 @@
+    print('Example: Output vertical grid mask for the C grid')
+    print('mit.hfac(depth,rF,0.3,50,"C")')
+
+    [hFacC]=mit.hfac(depth,rF,0.3,50,'C')
```
Should this be `bathy`?

I get an error when running this example.

dramauh replied:
I think i pressed something wrong. it says @dramauh
[Merge branch 'mitgcmutils_update' of](https://github.com/MITgcm/MITgcm/pull/727/commits/c1ab038633bd0c7219bd2652d60f3c30cdc512e1) https://github.com/dramauh/MITgcm. Can someone delete that part, please?

edoddridge reviewed (commented):
I've had a go at running the examples hit an error in the `hfac` one - I think it's a simple fix though.

@jahn makes a good point about mixing up the imports, but I don't know if or how we want to fix that. Perhaps it's fine to assume that anyone running this has matplotlib installed?

edoddridge commented:
> I think i pressed something wrong. it says @dramauh
[Merge branch 'mitgcmutils_update' of](https://github.com/MITgcm/MITgcm/pull/727/commits/c1ab038633bd0c7219bd2652d60f3c30cdc512e1) https://github.com/dramauh/MITgcm. Can someone delete that part, please?

I can force push a change to remove that commit from the branch but it will cause havoc for you. You would need to delete your local copy of the branch and download this copy in its place. I don't think it is worth it. Instead, we can squash merge and that commit won't matter.

I'm happy with this PR now.

@jahn - are the additional dependencies a deal breaker for you?

edoddridge commented on utils/python/MITgcmutils/MITgcmutils/density.py:
```diff
@@ -0,0 +1,1063 @@
+    Notes
+    -----
+    AUTHOR:     Martin Losch 2002-08-09  (mlosch@mit.edu)
+
```
I still think it's worth including the attribution that it was converted to python by @jahn that was in the original `jmd95.py` file.

"# converted to python by jahn on 2010-04-29"

dramauh replied:
I will add attribution done by Jahn. About the second issue. I think you have an outdated version. The last version does not have more than 780 lines. Still I found there are several comments that point Martin as author, I will delete them.

edoddridge reviewed (changes requested):
Sorry to renege so quickly on that approval. I realised that there were a couple of previous comments I had left that haven't been resolved yet.

dramauh commented:
I changed the Martin attributions in the comments. Also, I was not sure who and when wrote TEOS10 in MITgcm, so I did not write anything. If anyone knows who and when did it, I can update it. May you @edoddridge take a look, please? I can still improve the comments.

jrscott commented:
@dramauh circling back here to try and get this merged finally

I show two issues:
- the thread says @edoddridge has requested changes that haven't been addressed, but going through I think you might have already done so (?) Can you go through the thread and resolve all the conversations that show as open?
- Can you see if you can resolve the conflicts? Or advise us if you need help here?

@edoddridge can you do a quick check that you are happy in this final form? 

Thanks!

dramauh commented:
Hi,
I corrected them. I requested another review time ago

edoddridge commented:
Sorry for the delay. I missed the original notification. There are some conflicts to resolve, but I'm happy with the changes.

jrscott commented:
thanks @edoddridge 
@dramauh @mjlosch can you resolve the open conversations (either click resolve if completed and/or verify the changes were made) and we can finally merge this PR? thx

mjlosch commented:
I resolved the remaining conflicts, and conversations. 
@dramauh It would be cool if you could quickly check that I didn't break anything.

dramauh commented:
Hi,

I have just taken a look to the changes, there should not be a problem. I see there is a compilation error in the latex, but it is not related to the python package

mjlosch commented:
That's odd, I can compile the latexpdf target without problems.
And yes, the error message of the CI:
```
! Extra }, or forgotten \endgroup.
\sbox  ...box {\color@setgroup #2\color@endgroup }
                                                  
l.1998 ...ion}{note}{Pressure method calling tree}
                                                  
!  ==> Fatal error occurred, no output PDF file produced!
```
does not point to any python related stuff. @jahn, @jm-c any idea?

jahn commented:
Looks like the latex in the github runner has a problem with admonitions now, master has the same problem.  I could generate the pdf for this PR just fine, so I guess we can ignore it.

jahn commented:
Looks like this has been resolved. I've rerun the failed test.

jm-c commented:
It seems to me that this PR is ready to be merged, @dramauh am I correct ? If so and if no one wants to add things here I could merge it soon after.

jm-c commented:
Reading more carefully the past comments, It's seems it's ready to be merged. Will do that later today.
